### PR TITLE
Patient bulk upload email notification

### PR DESCRIPTION
**Pull request by georgehager**
# BACKEND PULL REQUEST

## Related Issue #4440 

Adds an email notification once patient bulk upload is complete. The email includes a link back to the application, where if logged in, the user can view the newly uploaded patients.

⚠️ **Note that this is built on top of branch `emma/async-3`. and cannot be merged until that branch is merged. I'll leave this PR as a draft until the async behavior for patient bulk upload is merged into main.**

💡 **[This compare link](https://github.com/CDCgov/prime-simplereport/compare/emma/async-3...george/4440-patient-upload-email) will probably be more useful to view changes**

## Changes Proposed

-  adds an email template for patient bulk upload
- when save to the database save in the async method for patient bulk upload is complete, an email is sent to notify the user
- if processing fails, an email is sent to notify the user (screenshot [here](https://github.com/CDCgov/prime-simplereport/issues/4440#issuecomment-1324262939))

## Additional Information

- I added one unit test for the new email template
- There should probably be another test to verify that the success email gets sent in the async method, however this will be difficult to write until the changes for tests are completed on `emma/async-3`. I'll update this when we have a clearer testing strategy for the async method. Backend tests are expected to fail right now.
- I'll also need to to add a unit test for the error email behavior, if processing within the while loop fails, or if the saveAll to the db fails.

Here is the sequence of events: before patient upload, after patient upload, and then the email that gets sent:
- before
dev4 b4
- after
dev4 after
- email
dev4 email

When an error occurs in processing, sending the error email is triggered. This isn't manually testable in an environment, so here are the logs when I triggered errors in the code:
- processing error in while loop (turned off validation in `PatientBulkUploadService` and uploaded an invalid csv
Screen Shot 2022-11-22 at 4 01 21 PM
- processing error in saveAll (forced an error in the `addPatientsAndPhoneNumbers` method)
Screen Shot 2022-11-22 at 4 15 31 PM

## Testing

How should reviewers verify this PR?
- Go to dev4, upload patients via csv, and confirm that you receive an email notifying you when the upload is complete. Click the `See patients` link, which should take you directly to the patients page, if you are logged in to dev4.

## Checklist for Primary Reviewer
- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke tested
- [x] Any content updates (user-facing error messages, etc) have been approved by content team
- [ ] Any changes that might generate questions in the support inbox have been flagged to the support team
- [ ] Changes comply with the SimpleReport Style Guide

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        }
+      } catch (IllegalArgumentException e) {
+        String errorMessage = "Error uploading patient roster";
+        log.error(
+            errorMessage
+                + " for organization "
+                + currentOrganization.getExternalId()
+                + " and facility "
+                + facilityId);
+        throw new IllegalArgumentException(errorMessage);
```
Comment (BobanL) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:145]: Wouldn't we also want to send an email if the upload failed?

**Review by BobanL — COMMENTED** [review]
Tested it and it looks good :) just one question

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java (hunk 1 above)
Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:145]: @BobanL that is an excellent question. Let me check in with design about that.

Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:145]: +1 to this, I think we'd want some kind of "try again or contact support" messaging.

Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:149]: @BobanL @emmastephenson please take a look at the new changes and let me know if the error handling makes sense to you, specifically in this area:
https://github.com/CDCgov/prime-simplereport/blob/7c2ecfa52a74f6f4845474e4458d54b5f6cd83ff/backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java#L153-L153

Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:149]: Note that it's really not possible to force this error in an environment, so I can't really have you test it manually! But once Emma's async unit test is ready, I'll add on to that.

**Review by BobanL — COMMENTED** [review]
Changes with the error email look good to me!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+          errorMessage
+              + " for organization "
+              + currentOrganization.getExternalId()
+              + " and facility "
+              + facilityId);
+      throw new IllegalArgumentException(errorMessage);
+    }
+  }
+
+  private void sendErrorEmail(
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:170]: There's a lot of duplicate code between this and the `sendSuccessEmail` method - I wonder if you could consolidate down to a single `sendEmail` method and pass a boolean for success/failure or something like that. Up to you though, this way may be more readable.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          phoneNumbersList.addAll(newPhoneNumbers);
+          newPhoneNumbers.stream().findFirst().ifPresent(newPatient::setPrimaryPhone);
+
+          patientsList.add(newPatient);
+        }
+      } catch (IllegalArgumentException e) {
+        sendErrorEmail(uploaderEmail, simplereportUrl, currentOrganization);
+        String errorMessage = "Error uploading patient roster";
+        log.error(
+            errorMessage
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:144]: This block of logging code is duplicated in the next try/catch for database saves - could you extract it out into a `logFailure` method or similar?

**Review by emmastephenson — COMMENTED** [review]
Looks great! Just a quick comment about some code deduplication.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java (hunk 2 above)
Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java]: ✅

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java (hunk 3 above)
Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:149]: ✅

**Comment by georgehager** [comment]
Closed in favor of #4852, which is just a new branch with the same changes to keep things tidy.

**Pull request closed without merging**